### DISCUSSION TITLE
Create Sinusoidal grating using shaders with inter-trial intervals

### DISCUSSION THREAD

**Discussion by ThetaS** [bonsai-rx/bonsai#1038] [Q&A] [answered]
Hi all,

I was looking at the Shader tutorial with Bonsai from the old google group: https://groups.google.com/g/bonsai-users/c/97lvnMhAoPM/m/GqHJ7mlzBAAJ . I tried to run the code example provided but it seems to be an older version and I got an error when attempting to update it. I wonder how merge the shader-created sinusoidal grating to normal Bonsai or Bonvision processes? 

As a side note, I need to create simulus from custom shaders because I want to crop one part of the grating to display other things (like a pop-out). When I overlay a small grating aperture onto another big one, the small one in the front would show its bounding box (a square). Does anyone have any idea of solving this issue with Bonvision? Thank you.

**Comment by bruno-f-cruz** [comment]
To draw sin gratings you can use the DrawGratings node. There are several posts in the repo that you might want to look at.
Regarding the second question, I am not sure I followed :( . Any chance you can draw a mockup of what you are looking for?

**Comment by ThetaS** [comment, marked as answer]
Hi Bruno,

Thanks for the reply. I've though of using DrawGrating node but in previous version of Bonvision, overlaying a stimulus onto another would appear a black square bounding box. It seems like you've fixed this problem already. But I think it would be great to cover how shaders can be loaded in Bonsai in this thread. Bonsai has been updated very much since the "Shader tutorial" has appeared in the google group, especially the DrawMesh node. Could you briefly mention how to load shader files in "UpdateFrame" and use it in "DrawMesh"? Thank you!


![center_surround](https://user-images.githubusercontent.com/32575952/190020043-3d1c4995-f4b7-4060-bfb1-9c72e19973e0.png)

**Reply by bruno-f-cruz** [reply]
The "old" way of drawing the gratings is still largely used. The difference is that it has been wrapped inside a BonVision node for everyone's convenience (+ some new parameterization). You can still see it by un-grouping the gratings node :). See attached.

Regarding a brief explanation. We have been discussing writing down a small tutorial on how to use the shader s package in Bonsai. The problem is that it is a large package and we have not quite had the time to tackle it.

In very broad strokes, the general steps to use a shader are:

- Load it as a resource (check `BonVisionResources `for an example of how to do this). 
- Next, you latch on to an update event (`UpdateFrame `or `RenderFrame`) and update all the parameters of your shader (`UpdateUniform`)
- Finally, run it on a pre-loaded mesh (for most 2d stimuli a simple normalized Quad will do the trick) with `DrawMesh`.

Cheers,
B

[Gratings.zip](https://github.com/bonsai-rx/bonsai/files/9564087/Gratings.zip)
